qsimcirq/qsim_simulator.py

```diff
@@ -376,27 +376,34 @@ def _sample_measure_results(
                 seed=self._prng,
             )
 
-            for i in range(repetitions):
-                for key, op in meas_ops.items():
-                    meas_indices = [qubit_map[qubit] for qubit in op.qubits]
-                    invert_mask = op.gate.full_invert_mask()
-                    for j, q in enumerate(meas_indices):
-                        results[key][i][j] = full_results[i][q] ^ invert_mask[j]
+            for key, op in meas_ops.items():
+                meas_indices = [qubit_map[qubit] for qubit in op.qubits]
+                invert_mask = op.gate.full_invert_mask()
+                # Match result order to order in ops, then apply invert mask
+                permuted_results = full_results[:, meas_indices]
+                results[key] = np.logical_xor(permuted_results, invert_mask)
+
         else:
             options["c"] = self._translate_circuit(
                 program,
                 translator_fn_name,
                 cirq.QubitOrder.DEFAULT,
             )
+            measurements = np.empty(
+                shape=(
+                    repetitions,
+                    sum(cirq.num_qubits(op) for op in meas_ops.values()),
+                )
+            )
             for i in range(repetitions):
                 options["s"] = self.get_seed()
-                measurements = sampler_fn(options)
-                for key, bound in bounds.items():
-                    invert_mask = meas_ops[key].gate.full_invert_mask()
-                    for j in range(bound[1] - bound[0]):
-                        results[key][i][j] = int(
-                            measurements[bound[0] + j] ^ invert_mask[j]
-                        )
+                measurements[i] = sampler_fn(options)
+
+            for key, bound in bounds.items():
+                invert_mask = meas_ops[key].gate.full_invert_mask()
+                permutation = list(range(bound[0], bound[1]))
+                permuted_measurements = measurements[:, permutation]
+                results[key] = np.logical_xor(permuted_measurements, invert_mask)
```

Comment:
Here, too, the `^` operator should work on arrays:
```suggestion
                results[key] = permuted_measurements ^ invert_mask
```
Also, I think "permuted` is a bit misleading since we don't change the order, just extract a slice of qubits with the same order, so perhaps could just inline the slice expression here to avoid a potentially misleading name:
```suggestion
                results[key] = measurements[:, start:end] ^ invert_mask
```

Reply:
Applied the slice but kept `np.logical_xor`, same as above.